skyportal/handlers/api/plan.py

```diff
@@ -0,0 +1,478 @@
+
+from sqlalchemy.orm.exc import NoResultFound
+from sqlalchemy.orm import joinedload
+from astropy import time
+import ephem
+import gwemopt
+import numpy as np
+
+from baselayer.app.access import auth_or_token
+from ..base import BaseHandler
+from ...models import (
+    DBSession,
+    Instrument,
+    Localization,
+    ObservingPlan,
+    PlannedObservation,
+    Telescope,
+)
+
+plan_args_default = {
+    'filt': ['g', 'r', 'g'],
+    'exposuretimes': [300.0, 300.0, 300.0],
+    'doReferences': False,
+    'doUsePrimary': True,
+    'doBalanceExposure': False,
+    'doDither': False,
+    'usePrevious': False,
+    'doCompletedObservations': False,
+    'doPlannedObservations': False,
+    'cobs': [None, None],
+    'schedule_type': 'greedy',
+    'filterScheduleType': 'block',
+    'airmass': 2.5,
+    'schedule_strategy': 'tiling',
+    'mindiff': 30.0 * 60.0,
+    'doMaxTiles': False,
+    'max_nb_tiles': 1000,
+    'doRASlice': True,
+    'raslice': [0.0, 24.0],
+}
+
+
+class PlanHandler(BaseHandler):
+    @auth_or_token
+    def post(self):
+        """
+        ---
+        description: Create an observation plan
+        tags:
+          - observingplans
+        requestBody:
+          content:
+            application/json:
+              schema: PlanHandlerPost
+        responses:
+          200:
+            content:
+              application/json:
+                schema: Success
+          400:
+            content:
+              application/json:
+                schema: Error
+        """
+
+        data = self.get_json()
+
+        telescope_name = data.get('telescope_name')
+        instrument_name = data.get('instrument_name')
+        dateobs = data.get('dateobs')
+        localization_name = data.get('localization_name')
+
+        telescope = (
+            Telescope.query_records_accessible_by(
+                self.current_user,
+            )
+            .filter(
+                Telescope.name == telescope_name,
+            )
+            .first()
+        )
+        if telescope is None:
+            self.error(message=f"Missing telescope {telescope_name}")
+
+        instrument = (
+            Instrument.query_records_accessible_by(
+                self.current_user,
+            )
+            .filter(
+                Instrument.telescope == telescope,
+                Instrument.name == instrument_name,
+            )
+            .first()
+        )
+        if instrument is None:
+            self.error(message=f"Missing instrument {instrument_name}")
+
+        localization = (
+            Localization.query_records_accessible_by(self.current_user)
+            .filter(
+                Localization.dateobs == dateobs,
+                Localization.localization_name == localization_name,
+            )
+            .first()
+        )
+        if localization is None:
+            self.error(message=f"Missing localization {localization_name}")
+
+        plan_args = copy.deepcopy(plan_args_default)
```

Comment:
What is this needed for?

Reply:
So when down the line, we make the parameters configurable through the interface, some of these arguments will change, but maybe a copy isnt necessary? I just get nervous whenever I see something like 
```plan_args = plan_args_default''''
and similar, but maybe it is fine here?

Reply:
If you are going to modify `plan_args` it makes sense to do a copy.  Deepcopy is usually reserved for instances where you have objects inside the dict that also need to be copied carefully.

Also note this handy new syntax:

```
In [1]: d = {'a': 4, 'b': 6}

In [2]: b = {**d}

In [3]: b['a'] = 5

In [4]: d
Out[4]: {'a': 4, 'b': 6}

In [5]: b
Out[5]: {'a': 5, 'b': 6}
```

Reply:
Let me try and be a bit more clear: the deepcopy handles the deeper nested dictionaries, so you don't run into this issue:

```
In [12]: d = {'a': 4, 'b': 6}

In [13]: d = {'a': 4, 'b': 6, 'c': {'foo': 'bar'}}

In [14]: e = {**d}

In [15]: e['c']['foo'] = 'something'

In [16]: d
Out[16]: {'a': 4, 'b': 6, 'c': {'foo': 'something'}}

```

Reply:
That first example was terrifying ;)